cmd/snap/cmd_run.go

```diff
@@ -118,6 +118,37 @@ func getSnapInfo(snapName string, revision snap.Revision) (*snap.Info, error) {
 	return info, nil
 }
 
+func createOrUpdateUserDataSymlink(info *snap.Info, usr *user.User) error {
+	// 'current' symlink for user data (SNAP_USER_DATA)
+	userData := info.UserDataDir(usr.HomeDir)
+	wantedSymlinkValue := filepath.Base(userData)
+	currentActiveSymlink := filepath.Join(userData, "..", "current")
+
+	var err error
+	var currentSymlinkValue string
+	for i := 0; i < 5; i++ {
+		currentSymlinkValue, err = os.Readlink(currentActiveSymlink)
+		// failure other than non-existing symlink is fatal
+		if err != nil && !os.IsNotExist(err) {
+			// TRANSLATORS: %v the error message
+			return fmt.Errorf(i18n.G("cannot read symlink: %v"), err)
+		}
+		if currentSymlinkValue == wantedSymlinkValue {
+			break
+		}
+		if !os.IsNotExist(err) {
+			_ = os.Remove(currentActiveSymlink)
+		}
+		if err = os.Symlink(wantedSymlinkValue, currentActiveSymlink); err == nil {
+			break
```

Comment:
If the error is EEXISTS let's do exactly as what we do for ENOENT above (abort and try again)